Chameleon clothing

Macoron opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Requires https://github.com/space-wizards/RobustToolbox/pull/2868

Add chameleon clothing for each inventory slot. It copies name, description and all sprites by visualizer client-side. Selection done by bound user interface.

It's added in uplink as chameleon backpack and full set of clothes (except belt and id-pda). The cost is 4 TC.

By default you can mimic ANY clothing in codebase. If you want to exclude some clothes use `IgnoreChameleon` tag. 

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->

https://user-images.githubusercontent.com/6161335/170279680-bb6b70e9-3705-403c-a62f-abf2031c029a.mp4


**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: Added chameleon clothing that can mimic any uniform. Watch out for suspicious syndicate imposters.

Macoron commented:
Oh, forgot requires #2868

Just-a-Unity-Dev commented:
this is gonna be so cool once identity gets added, stealth ops will be viable

moonheart08 reviewed (changes requested):
Upon in-game testing, the chameleon UI doesn't appear to be working at the moment? Not sure why.

Macoron commented:
Ups, I broke the thing while was trying to pass tests. Should be fixed now.

metalgearsloth commented on Content.Client/Clothing/UI/ChameleonBoundUserInterface.cs:
```diff
@@ -0,0 +1,54 @@
+        if (state is not ChameleonBoundUserInterfaceState st)
+            return;
+
+        var chameleonSystem = EntitySystem.Get<ChameleonClothingSystem>();
```
Don't use EntitySystem.Get<T> anymore.

Macoron replied:
Sadly, bound user interfaces can't inject systems :(

metalgearsloth commented on Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+    /// </summary>
+    public List<string> GetValidTargets(SlotFlags slot)
+    {
+        var list = new List<string>();
+        foreach (var availableSlot in _data.Keys)
+        {
+            if (slot.HasFlag(availableSlot))
+            {
+                list.AddRange(_data[availableSlot]);
+            }
+        }
+
+        // remove duplicates because some clothing can have multiple slots
+        return list.Distinct().ToList();
+    }
```
Just use a hashset.

metalgearsloth commented on Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+        return list.Distinct().ToList();
+    }
+
+    private void PrepareAllVariants()
+    {
+        var prototypes = _proto.EnumeratePrototypes<EntityPrototype>();
+
+        foreach (var proto in prototypes)
+        {
+            // check if this is valid clothing
+            if (!IsValidTarget(proto))
+                continue;
+            if (!proto.TryGetComponent(out ClothingComponent? item, _factory))
+                continue;
+
+            // sort item by their slot flags
+            // one item can be placed in several buckets
+            foreach (var slot in Slots)
+            {
+                if (!item.SlotFlags.HasFlag(slot))
+                    continue;
+
+                if (!_data.ContainsKey(slot))
+                {
+                    _data.Add(slot, new List<string>());
+                }
+                _data[slot].Add(proto.ID);
+            }
+        }
+    }
```
Doesn't handle prototype reloading.

metalgearsloth commented on Content.Client/Clothing/UI/ChameleonMenu.xaml.cs:
```diff
@@ -0,0 +1,86 @@
+        ClearGrid();
+
+        var group = new ButtonGroup();
+        var spriteSys = EntitySystem.Get<SpriteSystem>();
```
As above.

metalgearsloth commented on Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /// <summary>
+    ///     Get a list of valid chameleon targets for this slots.
```
```suggestion
    ///     Get a list of valid chameleon targets for these slots.
```

metalgearsloth commented on Content.Client/Items/Systems/ItemSystem.cs:
```diff
@@ -100,5 +100,26 @@ private bool TryGetDefaultVisuals(EntityUid uid, SharedItemComponent item, strin
+        // TODO: beware, there is EquippedPrefix setter which marks entity dirty on set
+        // so it resync data from server and drop override from chameleon visualizer
+        // right now I set it directly bypassing setter
+        // this probably will broke something latter, but please just remove setter from clothing
```
```suggestion
        // this probably will break something later, but please just remove setter from clothing
```

metalgearsloth commented on Content.Shared/Clothing/Components/ChameleonClothingComponent.cs:
```diff
@@ -0,0 +1,65 @@
+    /// </summary>
+    [ViewVariables(VVAccess.ReadOnly)]
+    [DataField("default", required: true, customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
+    public string SelectedId = "";
```
```suggestion
    public string? SelectedId;
```

metalgearsloth reviewed (changes requested):
I'm not a huge fan of using a blacklist for it.

metalgearsloth commented:
This is like 9/10ths of the way there so would be good to get it ingame when you or someone else has time.

Macoron commented:
I updated this to the latest master, addressed @metalgearsloth and rewrite visualizer logic. Now it used component state which is working much better (no more conflicts when item got dirty and was rewriten by server). Should be good for review.

metalgearsloth commented on Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:
```diff
@@ -0,0 +1,114 @@
+        return set;
+    }
+
+    private void PrepareAllVariants()
+    {
+        var prototypes = _proto.EnumeratePrototypes<EntityPrototype>();
+
+        foreach (var proto in prototypes)
```
Need to clear existing data on prototype reload

metalgearsloth commented on Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs:
```diff
@@ -0,0 +1,67 @@
+        if (proto.Abstract || proto.NoSpawn)
+            return false;
+
+        // check if it isn't marked as invalid chameleon target
+        if (proto.TryGetComponent(out TagComponent? tags, _factory) && tags.Tags.Contains("IgnoreChameleon"))
+            return false;
```
Tag needs to be whitelist instead of blacklist.

moonheart08 replied:
You sure? I think blacklist is really fine here as 99% of clothing should be valid, that's the point of the suit.

Macoron replied:
Yeah, whitelist will be real pain, because there are hundreds of potential targets that chameleon can mimic.

metalgearsloth replied:
> You sure? I think blacklist is really fine here as 99% of clothing should be valid, that's the point of the suit.

I checked with Remie at the time and we both agreed on a whitelist.

The pain is every contributor is going to have issues if they add something like say, clothingcomponent to a mob and now they can show that mob in the slot, because there's implicit behaviour now being done on another system you may not be aware of.

Macoron replied:
> I doubt you need more than like... 20 abstract prototypes tagged?

Tags doesn't inherit well. So if any clothing has some additional tag, you need to go explicitly to their child and set `ChameleonWhitelist` tag. Or it will be overwritten by newly added tags.

> add something like say, clothingcomponent to a mob

Well, the actual problem is that right now codebase doesn't distinguish entities that are "fitting into inventory slot" and entities that are actual clothing. For example fireaxe or guitar are also consider to be clothing when they are clearly not.

I can suggest other solution. I can set whitelist of prototype ids explicitly in some `ChameleonWhitelistPrototype` and save id of this prototype in component. This will give explicit control over what chameleon clothing can mimic, but also add a lot of boilerplate.

metalgearsloth replied:
I think just putting the whitelist tag on all of the base clothing prototypes OR the `Clothing` prototype should probably work as stuff that shouldn't show up since rats don't inherit from it anyway?

I had a look and there's only 7 prototypes in the clothing folder with the Tag component on them.

Macoron replied:
I added `WhitelistChameleon` to `Clothing` prototype. Which means that almost every clothing is can be target for chameleon again. If `Clothing` child don't want to be chameleon target, usually it should something like this.

```
    - type: Tag
      tags: [] # ignore "WhitelistChameleon" tag
```

mirrorcult replied:
if we go ahead with not using tagcomp for most stuff then this is just going to become even more hell. as it stands right now this basically -is- a blacklist because stuff needs to specify the empty tag anyway and it automatically inherits the whitelist. theres literally no point to making it a whitelist @metalgearsloth

mirrorcult replied:
if you want to avoid the 'clothing on mob' thing and make it not CBT with inheritance you could just make this a field on clothingcomponent i guess?

metalgearsloth replied:
> not to mention that now clothing which specifies other tags will be unavailable for chameleon unless people remember to put it there

Blacklist has meta issues:
- forbidden items get included (e.g. cat ears)
- Unremoveable items like hardsuit helmets get included

Maybe it should just clone like the agent card because I foresee github issues being made and it's gonna get the same meta abuse that centcomm uniforms did on uplink before we removed them (which are also included here).

Also unrelated but the UI doesn't align with the actual sprite you select.
![image](https://user-images.githubusercontent.com/31366439/187674800-1e18efbf-8163-4404-9e68-52f2b07aee51.png)

Plus the mask being pulled up / down state isn't synced.

Macoron replied:
> if you want to avoid the 'clothing on mob' thing and make it not CBT with inheritance you could just make this a field on clothingcomponent

I don't think that it should be done this way. Some forks may not use chameleon clothing. So it will be useless field that breaks muh SOLID

I agree that current solution with tag overriding is total hell. Of course I can go and mark with tag every single clothing prototype child directly (not their parents), but this is even worse. Because if parent want to add some tag, you need to go for each child proto and add it too with all previously inherited tags manually.

Macoron replied:
> forbidden items get included (e.g. cat ears)

Intended, unless it marked as blacklist for chameleon directly

> Unremoveable items like hardsuit helmets get included

Well, right now they are marked as no spawn, so they don't appear in search. But anyway they doesn't have unremovable behaviour, chameleon just mimic their appearance. Can't see why would it be an issue.

> Maybe it should just clone like the agent card because I foresee github issues being made and it's gonna get the same meta abuse that centcomm uniforms did on uplink before we removed them (which are also included here).

Before identity system and crew manifest that would be a problem. Right now not.

I can add blacklist for chameleon clothing as separate prototype and don't mess with proto inheritance or tag at all. It will be just a list of all blacklisted or whitelisted prototypes.

Macoron replied:
Any ideas where should I move this whitelist issue? Current options are:
- Return it to blacklist tag
- Use current implementation with whitelist tag
- Move whitelist to separate prototype and set all clothing entities manually

metalgearsloth reviewed (changes requested):
Rest LGTM, good job!

Macoron commented:
> Also unrelated but the UI doesn't align with the actual sprite you select.

Fixed. Was wrong order of sprite asign for some prototypes. Should work with any now.

> Plus the mask being pulled up / down state isn't synced.

Any steps to reproduce? Works same way as regular mask.

https://user-images.githubusercontent.com/6161335/187686836-062b0a63-bd7f-4b78-bd9b-374c40cb65d9.mov

metalgearsloth commented:
> Any steps to reproduce? Works same way as regular mask.

If you have it pulled down and chameleon it appears pulled up.

Macoron commented:
> > Any steps to reproduce? Works same way as regular mask.
> 
> If you have it pulled down and chameleon it appears pulled up.

Fixed by making this chameleon mask no longer a mask. Holographic projection can't transmit air, sorry.

Talking seriously, I can fix that by doing some event checks, but execution order on client and server a bit weird. I can take a look at that latter.

moocowswag commented:
Probably irrelevant but i wanted to say that tg cameleon clothing had slight armor values so the user would be tankier than they appear. Not arguing for it to have armor just saying that was a thing.
